Support history for specific ratingKeys

zSeriesGuy opened the pull request:
This change adds history selection for specific ratingKeys, much like the Plex "View Play History".
History can be requested from a movie, show, season, episode, artist, album, or track.

- Add history method to PlexPartialObject
- Add ratingKey arg to server history method.
- Add tests for movie, show, season, episode, artist, album, and track history.

I added tests. However, I have not actually run them. I had a separate test against my own server. Hopefully I created these tests correctly.

pkkid commented:
This looks good to me. I'm going to leave it up for a day or two to let others comment or review before merging. Thanks.

zSeriesGuy commented:
I see audio does not have a markwatched/markunwatched. Removed those from the audio test.

zSeriesGuy commented:
Well, after that change to the tests, I see it is still marked failed.
The audio files are coming back with assert 0 because there is no play history for them and I don't know how they could be marked played.
I'll defer to you to decide how to code these tests.

blacktwin commented:
Update the docstring for the added `ratingKey`. 

For the history function in base.py shouldn't the `maxresults=9999999`, `mindate=None` be added as well? That way we can isolate the search history for the item?

Should `librarySectionID` attribute be added? Along with Users (`accountID`)? To match Plex's webui history area. If thats the case then history should be moved to myPlex.py instead of server.py, no? library.py and myPlex.py and server.py would need similar history functions as well. If we want to go that route. Or this can just be a stepping stone into the direction?

zSeriesGuy commented:
I verified that the returned items from history contains the correct accountID of the user that played the item. 

I noticed that the returned objects have an empty player list attribute. History returns a deviceID that could be used to look up a player device. I started to add a player class. But I noticed that client.py already defines a player object. However, returned clients are not the same as player devices. You can see all of the player devices with key /devices. The returned devices contain name, platform, createdAt attributess which is what the Plex list play history shows.

blacktwin commented:
```python
movie = plex.fetchItem(106740)
movie.reload()
pp.pprint(vars(movie.history()[0]))
```
```
{   '_data': <Element 'Video' at 0x00000000048FF420>,
    '_details_key': '/library/metadata/106740?checkFiles=1&includeExtras=1&includeRelated=1&includeOnDeck=1&includeChapters=1&includePopularLeaves=1&includeConcerts=1&includePreferences=1',
    '_initpath': u'/status/sessions/history/all?metadataItemID=106740&sort=viewedAt%3Adesc&X-Plex-Container-Size=50&X-Plex-Container-Start=0',
    '_server': <PlexServer:https://127.0.0.1:32>,
    'accountID': 1234455,
    'addedAt': None,
    'art': None,
    'audienceRating': None,
    'audienceRatingImage': None,
    'chapterSource': None,
    'chapters': [],
    'collections': [],
    'contentRating': None,
    'countries': [],
    'directors': [],
    'duration': None,
    'fields': [],
    'genres': [],
    'guid': None,
    'key': '/library/metadata/106740',
    'labels': [],
    'lastViewedAt': None,
    'librarySectionID': None,
    'listType': 'video',
    'media': [],
    'originalTitle': None,
    'originallyAvailableAt': datetime.datetime(2019, 5, 30, 0, 0),
    'players': [],
    'playlistItemID': None,
    'primaryExtraKey': None,
    'producers': [],
    'rating': None,
    'ratingImage': None,
    'ratingKey': None,
    'roles': [],
    'session': [],
    'sessionKey': None,
    'similar': [],
    'studio': None,
    'summary': None,
    'tagline': None,
    'thumb': '/library/metadata/106740/thumb/-1',
    'title': "The Movie",
    'titleSort': "The Movie",
    'transcodeSessions': [],
    'type': 'movie',
    'updatedAt': None,
    'userRating': None,
    'usernames': [],
    'viewCount': 0,
    'viewOffset': 0,
    'viewedAt': datetime.datetime(2019, 9, 23, 15, 44, 9),
    'writers': [],
    'year': None}
```
There is still a lot of missing information that could be useful to have. 

Also what I was referring to before was to be able to use `history` for a user or library section object so `myUser.history()` or `movieSection.history()` and get the same results as found in Plex's webUI history area.

zSeriesGuy commented:
I added more history methods.
With these changes, you can now retrieve play history for:
- Your MyPlexAccount: returns list of all history for the owner for all sections on all servers.
- Your MyPlexAccount.library: returns list of all history for the owner for all sections on this server.
- Your MyPlexAccount.library.Section: returns list of all history for the owner for this section on this server.
- Your MyPlexServer (this is not new): returns list of all history for all on this server.
- User Account: Returns list of all history for this user on all shared plex server sections.
- User Server: Returns list of all history for this user for all shared sections on this server.
- User Section: Returns list of all history for this user for this shared section on this server.
- Movie: Returns play history for this movie.
- Show: Returns play history for this show.
- Artist: Returns play history for this Artist.
- Album: Returns play history for this Album.
- Track: Returns play history for this track.

I add a server method to MyPlexUser to return the named server, Returns a MyPlexServerShare.
Added a section and a sections method to return the sections in the MyPlexServerShare.

Please take a look at myplex.py history method in MyPlexServerShare, line 786.
Is there a better way to get a plexServer object?

zSeriesGuy commented:
I also created a test_history in the tests directory.
But I'm not sure how to affirm the tests for users. Some items would need to be marked watched.

blacktwin commented on plexapi/server.py:
```diff
@@ -322,7 +322,7 @@ def installUpdate(self):
             return self.query(part, method=self._session.put)
 
-    def history(self, maxresults=9999999, mindate=None, ratingKey=None):
+    def history(self, maxresults=9999999, mindate=None, ratingKey=None, accountID=None, librarySectionID=None):
```
`maxresults=9999999` causes the search to take a long time. Extending `maxresults=9999999` out to the other history functions would allow for the user to specify the `maxresults` ie: `user.history(maxresults=5)`. Unless there is a better way?

zSeriesGuy commented:
I agree. I added maxresults and mindate to the other history functions.
